File: Common/Model/JWRequest.m
//
//  JWRequest.m
//  BusRider
//
//  Created by John Wong on 12/16/14.
//  Copyright (c) 2014 John Wong. All rights reserved.
//

#import "JWRequest.h"
#import "STHTTPRequest.h"
#import "JWUserDefaultsUtil.h"
#import "JWCityItem.h"

#define JWDataErrorKey @"JWDataError"

//#define FAKE

#ifdef FAKE
#define kJWHost @"localhost:7000"
#else
#define kJWHost @"api.chelaile.net.cn:7000"
#endif


@interface JWRequest ()

@property (nonatomic, strong) STHTTPRequest *request;

@end


@implementation JWRequest

- (NSString *)actionName
{
    NSAssert(NO, @"override this method %s", __FUNCTION__);
    return nil;
}

- (NSDictionary *)params
{
    return nil;
}

- (void)loadWithCompletion:(JWCompletion)completion
{
    [self loadWithCompletion:completion progress:nil];
}

- (void)loadWithCompletion:(JWCompletion)completion progress:(JWProgress)progress
{
    NSString *checkResult = self.validateParams;
    if (checkResult) {
        NSError *error = [NSError errorWithDomain:JWDataErrorKey code:0 userInfo:@{
            NSLocalizedDescriptionKey : checkResult
        }];
        completion(nil, error);
        return;
    }
    NSLog(@"JWRequest: load %@", [self urlPath]);

    if (self.request) {
        [self.request cancel];
    }
    if (progress) {
        progress(20);
    }
    self.request = [STHTTPRequest requestWithURLString:[self urlPath]];
    self.request.completionBlock = ^(NSDictionary *headers, NSString *body) {
        NSString *jsonString = [[body stringByReplacingOccurrencesOfString:@"**YGKJ" withString:@""] stringByReplacingOccurrencesOfString:@"YGKJ##" withString:@""];
        NSError *error = nil;
        id dict = [NSJSONSerialization JSONObjectWithData:[jsonString dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:&error];
        if (error == nil && [dict isKindOfClass:[NSDictionary class]]) {
            NSDictionary *jsonr = dict[@"jsonr"];
            NSDictionary *infoDict = jsonr[@"data"];
            if (infoDict.count > 0) {
                NSLog(@"JWRequest: response %@", infoDict);
                completion(infoDict, nil);
            } else {
                error = [NSError errorWithDomain:JWDataErrorKey
                                            code:301
                                        userInfo:@{
                                            NSLocalizedDescriptionKey : jsonr[@"errmsg"]
                                        }];
                NSLog(@"JWRequest: error %@", error);
                completion(nil, error);
            }

        } else {
            NSLog(@"JWRequest: error %@", error);
            completion(nil, error);
        }
    };
    self.request.errorBlock = ^(NSError *error) {
        if (error.code == kSTHTTPRequestCancellationError ||
            error.code == -999) { // Connection was cancelled.
            return;
        }
        completion(nil, error);
    };
    self.request.downloadProgressBlock = ^(NSData *data, int64_t totalBytesReceived, int64_t totalBytesExpectedToReceive) {
        NSLog(@"JWRequest: progress %ld / %lld", (unsigned long)totalBytesReceived, totalBytesExpectedToReceive);
        if (progress) {
            CGFloat percent = totalBytesReceived * 100 / (totalBytesExpectedToReceive == NSURLResponseUnknownLength ? 10240 : totalBytesReceived);
            if (percent > 0.2) {
                progress(percent);
            }
        }
    };
    [self.request startAsynchronous];
}

- (NSString *)urlPath
{
    NSMutableString *paramString = [[NSMutableString alloc] init];
    NSDictionary *paramDict = self.params;
    for (NSString *key in paramDict) {
        [paramString appendFormat:@"&%@=%@", key, paramDict[key]];
    }
    NSString *cityId = @"";
    JWCityItem *cityItem = [JWUserDefaultsUtil cityItem];
    if (cityItem) {
        cityId = [NSString stringWithFormat:@"&cityId=%@", cityItem.cityId];
    }
    return [NSString stringWithFormat:@"http://%@/%@/%@.action?sign=&v=5.2.0&s=IOS&sv=9.1&vc=10070%@%@",
                                      kJWHost,
                                      [self isKindOfClass:NSClassFromString(@"JWCityRequest")] ? @"goocity" : @"bus",
                                      self.actionName,
                                      cityId,
                                      [paramString stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]]];
}

- (NSString *)validateParams
{
    return nil;
}

- (void)dealloc
{
    [_request cancel];
}

@end

File: BusRider WatchKit Extension/JWRequest.swift
//
//  JWRequest.swift
//  BusRider
//
//  Created by John Wong on 4/7/15.
//  Copyright (c) 2015 John Wong. All rights reserved.
//

import Foundation

class JWRequest: NSObject {
    var request:STHTTPRequest?
    
    internal override init() {
        super.init()
    }
    
    func params() -> Dictionary<String, AnyObject> {
        return Dictionary()
    }
    
    func actionName() -> String {
        return ""
    }
    
    internal func urlPath() -> String {
        var paramString = String()
        for (key, value) in self.params() {
            paramString = paramString.stringByAppendingString("&\(key)=\(value)")
        }
        var cityId = "";
        let cityItem = JWUserDefaultsUtil.cityItem()
        if  let cityItem = cityItem {
            cityId = String(format: "&cityId=%@", cityItem.cityId)
        }
        return String(format: "http://%@/%@/%@.action?s=IOS&v=2.9%@&sign=%@", AppConfiguration.host, self.isKindOfClass(JWCityRequest) ? "wow" : "bus", self.actionName(), cityId, paramString)
    }
    
    func loadWithCompletion(completion:(NSDictionary?, NSError?) -> Void) {
        request?.cancel();
        self.request = STHTTPRequest(URLString: self.urlPath())
        self.request?.completionBlock = {
            (headers: Dictionary!, body: String!) in
            let correctString = body as NSString
            let jsonString = correctString.stringByReplacingOccurrencesOfString("**YGKJ", withString: "").stringByReplacingOccurrencesOfString("YGKJ##", withString: "", options: NSStringCompareOptions(), range: nil)
            let jsonObject: AnyObject
            do {
                let data: NSData = jsonString.dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: false)!;
                jsonObject = try NSJSONSerialization.JSONObjectWithData(data, options: NSJSONReadingOptions())
                let dict = jsonObject as! NSDictionary
                let jsonr: NSDictionary = dict["jsonr"] as! NSDictionary
                let infoDict: NSDictionary = jsonr["data"] as! NSDictionary
                if infoDict.count > 0 {
                    completion(infoDict, nil);
                } else {
                    let error = NSError(
                        domain: AppConfiguration.dataErrorDomain,
                        code: 301,
                        userInfo: [
                            NSLocalizedDescriptionKey: jsonr
                        ])
                    completion(nil, error)
                }
            } catch let error as NSError {
                completion(nil, error);
            }
        }
        self.request?.errorBlock = {
            (error: NSError?) in
            if error?.code == 1 {
                return
            }
            completion(nil, error)
        }
        self.request?.startAsynchronous()
    }
}